Sources/System/FileOperations.swift

```diff
@@ -370,4 +370,52 @@ extension FileDescriptor {
   public func dup2() throws -> FileDescriptor {
     fatalError("Not implemented")
   }
+  
+  /// A pair of `FileDescriptor` values represening a pipe.
+  ///
+  /// You are responsible for managing the lifetime and validity
+  /// of the `read` and `write` `FileDescriptor` values,
+  /// which can be closed independently or via helper methods on `Pipe`, but not both.
+  public struct Pipe {
```

Comment:
Mentioned briefly in https://github.com/apple/swift-system/issues/57. I think the intention is to help avoid mixing up the read and write ends of the pipe, as tuple labels are not strongly enforced for destructuring assignment.

There's a point at which it's worth turning a labeled tuple into a proper nominal type, but I'm not sure where on the spectrum pipe falls. If it is usually the case that pipes are immediately split apart, then I think a tuple is fine. On the other hand, are both ends of a pipe typically closed together?

Reply:
One other consideration: if we change `FileDescriptor.closeAfter` to pass the file descriptor to the provided closure (as suggested in #54) then having similar behavior for `Pipe` would be nice (and require a dedicated `struct`). The main use case I see for closing both ends of a pipe at once is mentioned [here](https://github.com/apple/swift-system/pull/58#discussion_r683740434), namely forking off two processes you want connected with a pipe. This is important for shell-like and orchestration use cases.

Reply:
The thing is, mixing up the read and write ends of the pipe is a fairy cheap mistake to make, as mistakes go. Reading from or writing to the wrong end will immediately throw an error. 

I do agree it would be nice to prevent such mixups at compile time, but going with a struct makes structural decomposition a little more difficult, and (more importantly) it gives the (in my opinion) false impression that there is an API-level abstraction called `Pipe` that comes with its own operations. There are no operations on pipes -- there are only operations on its independent end points.